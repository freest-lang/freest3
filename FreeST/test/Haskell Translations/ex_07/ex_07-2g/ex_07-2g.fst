-- VII exercise 2g

data Nat = Zero | Succ Nat

pot : Nat -> Nat -> Nat
pot b e =
    case e of {
        Zero -> Succ Zero,
        Succ n -> mult b (pot b n)
    }

mult : Nat -> Nat -> Nat
mult x y =
    case y of {
        Zero -> Zero,
        Succ b -> add x (mult x b)
    }

add : Nat -> Nat -> Nat
add x y =
    case x of {
        Zero -> y,
        Succ z -> Succ (add z y)
    }

number2 : Nat
number2 = Succ (Succ Zero)

main : Nat
main = pot number2 number2
--result = Succ (Succ (Succ (Succ Zero)))
