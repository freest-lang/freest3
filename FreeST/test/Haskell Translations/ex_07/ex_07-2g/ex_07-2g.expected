Succ (Succ (Succ (Succ Zero)))